Add SMBus HAL, dependent on `embedded-hal-async::I2c`

tullom opened the pull request:
Add SMBus support in the form of a helper trait built on top of an embedded-hal-async I2C implementation.

Based on the [SMBus v3,3 spec](https://smbus.org/specs/).

tullom committed:
Add SMBus HAL, dependent on embedded-hal-async::I2c

RobertZ2011 commented on embedded-mcu-hal/src/smbus/bus/asynch.rs:
```diff
@@ -0,0 +1,743 @@
+            if use_pec {
+                let pec = Self::get_pec_calc();
+                if let Some(mut pec) = pec {
+                    pec.write_u8(address << 1);
```
Is it possible to rewrite these functions in terms of a generic buffer read/write functions. I was thinking something like `write_buffer(address: _ , reg: _, pec: bool, buffer: &[u8])`. Then `write_byte` would be implemented as write_buffer(address, reg, pec, &[byte])`. This would allow implementers to override the core read/write functionality in a single place.

tullom replied:
I noodled around with this for a while and could not find a nice way to calculate the PEC in a generic way. A PEC byte needs to be appended to the end of a write buffer, of which the size of a buffer is only known to the calling function.

Even worse, any `i2c::transaction()` call cannot append a PEC byte to a write operation in a generic way because `Operation::Write()`s contain immutable references to the underlying buffer. The slice of `Operation`s is mutable, but a new `Operation` cannot be added without having some sort of allocated collection. The user could pass in storage for the PEC byte but i think that would make the API arguments confusing.

I tried defining my own `Operation`, but this breaks interop with `i2c::transaction()`. I cannot construct the slice of `[i2c::Operation]` from `[smbus::Operation]` because I would need to allocate a collection to pass along to `i2c::transaction()`.

TBH, the user is not intended to override any of these methods. So long as the underlying I2C driver conforms to the `embedded_hal_async::i2c::I2c` trait, there is a guarantee that the default SMBus trait methods will work. I could make this whole trait a wrapper struct on a I2C driver, but I think it more ergonomic to tack a trait to existing I2C drivers.

tullom replied:
Finally coming back to this and i understand how this combo approach is good. Allows anyone to impl the trait themselves, while exposing a struct that takes in I2C and PEC objects for default implementations. I also have helper methods on the struct to deduplicate code.

felipebalbi commented on embedded-mcu-hal/src/smbus/bus/asynch.rs:
```diff
@@ -0,0 +1,743 @@
+///         }
+///     }
+///
+///     fn to_kind(kind: crate::smbus::bus::ErrorKind) -> Self {
```
Looks more like a from_kind

tullom replied:
Yeah makes sense. Renamed to from_kind.

kurtjd reviewed (dismissed):
I think Robert's comments are a good idea but otherwise looks good.

RobertZ2011 commented on embedded-mcu-hal/src/smbus/bus/asynch.rs:
```diff
@@ -0,0 +1,763 @@
+                    crate::smbus::bus::ErrorKind::Pec,
+                ))?;
+
+                pec.write_u8(address << 1);
```
You should be able to use the same `i2c::transaction`-based logic as write block so that the caller doesn't have to allocate memory for the PEC byte?

tullom replied:
Done.

RobertZ2011 commented on embedded-mcu-hal/src/smbus/bus/asynch.rs:
```diff
@@ -0,0 +1,763 @@
+        use_pec: bool,
+        read: &mut [u8],
+    ) -> impl core::future::Future<Output = Result<(), <Self as crate::smbus::bus::ErrorType>::Error>> {
+        async move {
```
Similar to `write_buf`, can you use transactions here so the caller doesn't have to provide space for the PEC byte?

tullom replied:
Done.

tullom committed:
Added unit tests

tullom committed:
Refactor SMBus async trait into Smbus + SwSmbusI2c

Split the async SMBus surface into an abstract protocol trait and a
concrete software implementation.

The `Smbus` trait now declares only the SMBus protocol operations plus
the PEC associated items (`PecCalc`, `get_pec_calc`, `check_pec`). All
default implementations that bit-bang the protocol on top of an I²C bus have moved into a new `SwSmbusI2c<I, P>` wrapper struct that owns an `embedded_hal_async::i2c::I2c` bus and delegates PEC calculator
construction to a new `PecProvider` trait via the `P` parameter. HALs
with a hardware SMBus peripheral can implement `Smbus` directly without inheriting any I²C-specific machinery.

`write_buf`, `read_buf`, and `write_read_buf` are no longer trait
methods; they are inherent helpers on `SwSmbusI2c` that the other
protocol methods reuse, eliminating duplicated PEC setup and I²C error
mapping across the byte/word/block operations. PEC calculator priming
with the write-address byte and the truncation of the finished hash to
a single byte are likewise factored into private helpers
(`pec_calc_with_write_addr`, `finalize_pec_byte`).

The blanket `impl<T: Smbus + ?Sized> Smbus for &mut T` forwarding impl
is preserved, and `SwSmbusI2c::Error` is `ErrorKind` directly so callers
do not need to define a wrapper error type. Tests are reworked to drive
`SwSmbusI2c<I2cMock, _>` through small `PecProvider` impls (`TestPec`,
`NoPec`).

Assisted-by: GitHub Copilot:claude-opus-4.7

tullom commented on embedded-mcu-hal/src/smbus/bus/asynch.rs:
```diff
@@ -0,0 +1,763 @@
+                let mut pec = Self::get_pec_calc().ok_or(<Self as crate::smbus::bus::ErrorType>::Error::to_kind(
+                    crate::smbus::bus::ErrorKind::Pec,
+                ))?;
+                pec.write_u8(address << 1);
```
Done, `read_buf` and `write_buf`

tullom commented on embedded-mcu-hal/src/smbus/bus/asynch.rs:
```diff
@@ -0,0 +1,1374 @@
+                        Operation::Write(&[register]),
+                        Operation::Write(&word.to_le_bytes()),
+                        Operation::Read(&mut buf),
+                    ],
```
Not true, subsequent `Operation`s of the same type do not trigger repeated Start

felipebalbi commented on embedded-mcu-hal/src/smbus/bus/asynch.rs:
```diff
@@ -0,0 +1,1374 @@
+        &mut self,
+        address: u8,
+        byte: u8,
+        use_pec: bool,
```
How about making the API explicit? That is, `send_byte()` and `send_byte_with_pec()`

tullom replied:
Done.

RobertZ2011 reviewed (dismissed):
Nothing to add on top of the existing comments.

kurtjd reviewed (dismissed):
Looks good (though Copilot suggestions seem like they might be valid).

tullom commented on embedded-mcu-hal/src/smbus/bus/asynch.rs:
```diff
@@ -0,0 +1,1374 @@
+                    ],
+                )
+                .await
+                .map_err(|e| crate::smbus::bus::ErrorKind::from(e.kind()))?;
```
Not true, subsequent Operations of the same type do not trigger repeated Start

tullom commented on embedded-mcu-hal/src/smbus/bus/asynch/mod.rs:
```diff
@@ -0,0 +1,1374 @@
+use core::marker::PhantomData;
+
+use crate::smbus::bus::Error as SMBusError;
+use embedded_hal_async::i2c::{Error as I2cError, I2c, Operation};
```
Not true, no warnings exist

tullom commented on embedded-mcu-hal/src/smbus/bus/asynch.rs:
```diff
@@ -0,0 +1,1374 @@
+//! This module defines the [`Smbus`] async controller trait describing the
+//! SMBus protocol surface. The trait declares the protocol-level
+//! operations as required methods plus two associated items â€”
+//! [`Smbus::PecCalc`] and [`Smbus::get_pec_calc`] â€” and a
```
Done.

tullom commented on embedded-mcu-hal/src/smbus/bus/asynch/mod.rs:
```diff
@@ -0,0 +1,1374 @@
+    /// low byte of `computed_pec` is used.
+    fn check_pec(received_pec: u8, computed_pec: u64) -> Result<(), <Self as crate::smbus::bus::ErrorType>::Error> {
+        computed_pec
+            .eq(&received_pec.into())
```
I think this is better behavior than masking off the high bits, as the hasher should only ever return a u8 PEC. If we got something else, that indicates something wrong happened with the PEC calculation and we should error out.

tullom committed:
Adress latest review comments.

- Renamed `to_kind` method to `from_kind` in the `Error` trait for clarity.
- Updated the `ErrorKind` enum to include a new variant `BlockSizeMismatch`.
- Made API more explicit by breaking out the pec and non pec SMBUS transactions into their own methods.

tullom commented on embedded-mcu-hal/src/smbus/bus/asynch.rs:
```diff
@@ -0,0 +1,1374 @@
+            let mut buf = [0u8; 3];
+            let mut pec = Self::pec_calc_with_write_addr(address)?;
+            pec.write_u8(register);
+            pec.write_u16(word);
```
Done.

tullom commented on embedded-mcu-hal/src/smbus/bus/asynch/mod.rs:
```diff
@@ -0,0 +1,1374 @@
+                        Operation::Read(&mut pec_buf),
+                    ],
+                )
+                .await
+                .map_err(|e| crate::smbus::bus::ErrorKind::from(e.kind()))?;
+            pec.write(&msg_size);
+            pec.write(data);
+            Self::check_pec(pec_buf[0], pec.finish())?;
+        } else {
+            self.i2c
+                .transaction(
+                    address,
+                    &mut [
+                        Operation::Write(&[register]),
+                        Operation::Read(&mut msg_size),
+                        Operation::Read(data),
+                    ],
+                )
+                .await
+                .map_err(|e| crate::smbus::bus::ErrorKind::from(e.kind()))?;
+        }
+        Ok(())
```
Done, added `BlockSizeMismatch` for this specific error case.

tullom commented on embedded-mcu-hal/src/smbus/bus/asynch/mod.rs:
```diff
@@ -0,0 +1,1374 @@
+        use_pec: bool,
+        read: &mut [u8],
+    ) -> Result<(), crate::smbus::bus::ErrorKind> {
+        if use_pec {
+            let mut pec = Self::pec_calc_with_write_addr(address)?;
+            self.i2c
+                .read(address, read)
+                .await
+                .map_err(|e| crate::smbus::bus::ErrorKind::from(e.kind()))?;
+            let (pec_byte, rest) = read.split_last().ok_or(crate::smbus::bus::ErrorKind::Pec)?;
+            pec.write(rest);
+            <Self as Smbus>::check_pec(*pec_byte, pec.finish())?;
+        } else {
```
Good catch, done.

tullom committed:
Move tests into standalone file

jerrysxie commented on embedded-mcu-hal/src/smbus/bus/asynch/mod.rs:
```diff
@@ -0,0 +1,932 @@
+    /// Compares a received PEC byte against a computed PEC value. Only the
+    /// low byte of `computed_pec` is used.
+    fn check_pec(received_pec: u8, computed_pec: u64) -> Result<(), <Self as crate::smbus::bus::ErrorType>::Error> {
+        computed_pec
```
computed_pec is u64 but received_pec is u8. We are converting received_pec into u64 to compare it with computed_pec. Is the hasher guaranteed to return zeros in for bit 63 to bit 8?

felipebalbi replied:
PEC is CRC8, no? Why `u64` at all?

tullom replied:
Its because I define that the CRC calculator must implement `core::hash::Hasher`, which outputs its final hash in the form of u64. Most hashing crates implement this trait. The hasher should be guaranteed to return zeros for bit 63 to bit 8, and i check for this in the implementation by promoting the u8 to a u64.

jerrysxie replied:
Reading the `hasher` traits, I don't believe there is any guarantee that bit 63 to bit 8 would be zeros. It seems to be hashing algorithm dependent. If we know it is always picking CRC8, then we know we would only get an 8-bit CRC. But the current trait does not seem to enforce it has to be CRC8.

tullom replied:
Instead of using the associated type, ive just baked [a CRC8 crate](https://crates.io/crates/smbus-pec) into the `SwSmbusI2c` struct. The trait is much more straightforward now and allows the implementer to define their own PEC calculations without having to define a concrete type that impls the `core::hasher::Hasher` trait.

jerrysxie commented on embedded-mcu-hal/src/smbus/bus/asynch/mod.rs:
```diff
@@ -0,0 +1,932 @@
+use core::hash::Hasher;
+use core::marker::PhantomData;
+
+use crate::smbus::bus::Error as SMBusError;
```
This import seems to be not needed now.

tullom replied:
It actually is needed for this line:
https://github.com/OpenDevicePartnership/embedded-mcu/pull/7/changes#diff-e28912d00e735f1ec57e7c90fc926b9887eefc013a240d8383eb414416054866R77

jerrysxie commented on embedded-mcu-hal/src/smbus/bus/asynch/mod.rs:
```diff
@@ -0,0 +1,932 @@
+        &mut self,
+        address: u8,
+        register: u8,
+        data: &mut [u8],
```
For SMBus varliable data length responses, is the assumption here that the buffer passed in is always big enough for the largest data payload? The client has to fulfill this guanrantee?

tullom replied:
For block reads, the size of the response is actually encoded in the response itself as byte 1 (zero indexed). The spec also defines an upper size limit of 64 bytes. So yes, the client needs to fulfill this guarantee. I2C::ErrorKind, which is encoded in an SMBus error, has an `Overrun` error kind that the client can use to indicate this error.

jerrysxie replied:
How does the calling client detect this error? The code will return `BlockSizeMismatch` without the size being expected. The calling client would not know what the current expected size would be. If this is an assumption of the API, then we should document.

```
async fn block_read(
        &mut self,
        address: u8,
        register: u8,
        data: &mut [u8],
    ) -> Result<(), crate::smbus::bus::ErrorKind> {
        if data.len() > crate::smbus::bus::MAX_BLOCK_SIZE {
            return Err(crate::smbus::bus::ErrorKind::TooLargeBlockTransaction);
        }
        let mut msg_size = [0u8];
        self.i2c
            .transaction(
                address,
                &mut [
                    Operation::Write(&[register]),
                    Operation::Read(&mut msg_size),
                    Operation::Read(data),
                ],
            )
            .await
            .map_err(|e| crate::smbus::bus::ErrorKind::from(e.kind()))?;
        if usize::from(msg_size[0]) != data.len() {
            return Err(crate::smbus::bus::ErrorKind::BlockSizeMismatch); <-----------------
        }
        Ok(())
    }
```

tullom replied:
makes sense, at least in the case of the BQ40Z50, the block size is documented in the data sheet so the calling client knows how large the block read should be. I think it could make sense to check if `data.len()` is smaller than `msg_size[0]` and return an error in that case. Or we can embed `msg_size[0]` in the `BlockSizeMismatch` error.

tullom replied:
I embedded msg_size[0] and the buffer length in the error, and updated comments to document that this is the behavior. Also updated the tests to check for these cases.

jerrysxie commented on embedded-mcu-hal/src/smbus/bus/asynch/mod.rs:
```diff
@@ -0,0 +1,932 @@
+        &mut self,
+        address: u8,
+        use_pec: bool,
+        operations: &mut [u8],
```
We should consider cases where a reference to an array of size 0 is passed in all of our API where it takes a reference to an array.

tullom replied:
Specifically for this method, `write_buf()`, it is not a public api method, but rather a convenience method, and every use case of it makes it impossible to pass in an array of size zero. As for the other methods that take in a reference to an array, the logic is correct according to the spec. A size 0 block write is valid in SMBus v3 and above

jerrysxie replied:
I was more concerned about whether passing in an array of size 0 would induce a panic or not. If you vet through all of cases where this might happen and it is okay, then we can resolve this comment.

As an aside, I thought only quick command are allowed to have no data. I don't believe we can do a block read or write with block length (byte 0) set to 0.

tullom replied:
<img width="494" height="106" alt="image" src="https://github.com/user-attachments/assets/ea136e5c-52cf-4160-a4b5-469e6bc226f5" />

As per v3.3 of the spec, section 6.5.7, pg 42

jerrysxie reviewed (changes requested):
Sorry, it has been a while since I last reviewed this. So this is basically a new review.

felipebalbi commented on embedded-mcu-hal/src/smbus/bus/asynch/mod.rs:
```diff
@@ -0,0 +1,932 @@
+    /// Send Byte.
+    async fn send_byte(&mut self, address: u8, byte: u8) -> Result<(), <Self as crate::smbus::bus::ErrorType>::Error>;
+
+    /// Send Byte with PEC.
+    async fn send_byte_with_pec(
+        &mut self,
+        address: u8,
+        byte: u8,
+    ) -> Result<(), <Self as crate::smbus::bus::ErrorType>::Error>;
```
Ugh, my bad, I should have suggested before that the `_with_pec()` variants can easily get a default implementation built on top of the non-pec variants.

tullom replied:
Ah, should i revert to the way it was previously? im already using a convenience method to make the code repeatable and i dont think its needed to revert.

tullom committed:
Update BlockSizeMismatch variant to include received byte count and expected buffer length

tullom committed:
Remove PecProvider trait from Smbus and use smbus_pec crate for SwSmbusI2c

jerrysxie reviewed (approved):
Doh, cargo-vet for this repo is currently broken. But that should not block this PR, issue: https://github.com/OpenDevicePartnership/embedded-mcu/issues/28
